Site Pages: implement the popup menu using EllipsisMenu and Redux

jsnajdr opened the pull request:
This PR started out as an attempt to remove usages of `lib/posts/actions` (the post list Flux store actions) from the Site Pages list. The Flux store is no longer used by anyone to read from (it's all Redux now) and the Flux actions only send the REST API requests and the update an unused Flux store. Redux is not updated at all.

The Site Pages list is the only remaining usage of the `updatePostStatus` HoC that adds capability to display UI like this to the `Page` component:

![page-update-post-status](https://user-images.githubusercontent.com/664258/37109432-c6531a9a-21ee-11e8-8782-afa89a074d2b.gif)

Here I do the following:
- start dispatching Redux actions instead of Flux ones
- remove the overlay UI from `updatePostStatus`: after migrating to Redux, notices are shown when a post status is succesfully changed
- remove the capability of `updatePostStatus` to maintain opened/closed state of the popover menu and start using `EllipsisMenu` for that
- merge the `updatePostStatus` HoC into the `Page` component. At this moment, only the `updatePostStatus` method remains

The result is that the Site Pages now behave exactly the same as Blog Posts. The post list was redesigned by @Automattic/tanooki (master thread is at p8F9tW-vh-p2) recently and Site Pages were out of scope of that project. Now, the Site Pages list still doesn't use `PostTypeList`, but the UI behavior is very close.

**How to test:**
1. Go to a site with a few pages
2. Open the popup menu for a page and click "Trash". Verify that the page was moved to trash and that a success notice is shown.
4. Open the list of trashed pages and try to restore one. Verify that the trashed page was moved to "Drafts" (the 100% right behavior would be to move it to "Published" -- that's a pre-existing bug I'd like to fix soon)
5. In the list of trashed pages, try to delete one permanently ("Delete" action). Verify that the permanent delete really happened.
6. In the list of "Drafts", click on the "Publish" action in the menu. Verify that the page was moved to "Published"
7. Verify that the UI behavior is the same as in Blog Posts page and there are no significant differences.

shaunandrews commented:
I feel like this is a step backwards. Lets start by looking at what the current experience is like for deleting a page:

![current experience](https://user-images.githubusercontent.com/191598/37111316-5b6afcb0-220d-11e8-9d58-38e15cf12455.gif)

Visual, this hasn't aged too well. But the experience is nice — you get immediate feedback that "something" is happening, and then an obvious way to undo it.

Now, lets look at what this PR does (and how Posts work currently):

![new experience](https://user-images.githubusercontent.com/191598/37111353-78cd1306-220d-11e8-8b61-2a366936186a.gif)

This feels like a regression. There's no immediate feedback that something is happening — just a disappearing row in the list. Then, after a few seconds of wondering if anything worked, you get a success notice. Its not terrible, but I think its worse than the current experience.

--

I'm pretty sure this has been brought up before (though I can't find a PR or Issue), and I worked up this suggestion:

![proposed experience](https://user-images.githubusercontent.com/191598/37111466-be2d3ba6-220d-11e8-83fd-dab8df157d2d.gif)

nylen commented:
FWIW, we're looking at adding an "Undo" to this global notice in #20663.  This will help with one of the concerns (an obvious way to undo the action you just took).

I think it would be good to proceed with the removal of unused code in this PR, as it is no longer up to our data standards and also makes the current experience consistent across pages and posts.  Then, leaving the in-progress / deleted item in the list should be a separate enhancement.

I'm also not sure about the use of the "notice" styles inline in the list.  From what I have seen, notices are usually actionable and more "global" to a page or at least a section.  Are there other places where we use these styles to indicate progress or inlined in part of a section like this?

shaunandrews commented:
> the current experience consistent across pages and posts.

As long as we improve this newly consistent experience, I guess thats ok. I just worry that it'll be a long time until that happens.

> Are there other places where we use these styles to indicate progress or inlined in part of a section like this?

`/plugins/` is a good example of where notices are used inline, though perhaps not "correctly." That being said, this would be a new pattern.

jsnajdr commented:
@shaunandrews I love the suggestion to give some visual feedback. It's not entirely trivial to implement it though, for various very technical reasons related to how the Posts Redux store is implemented. I think it warrants its own PR.

jsnajdr commented:
One more thing: implementing the notices as global ones (shown in the top right corner) would be much easier from the technical standpoint, although I understand that having them inline is a much neater UX.

The main reason for that is: after I trash a Page, it disappears from the "Published" list and moves to the "Trash" one. But the proposal keeps a "shadow" page item in the "Published" list and show a notice with "Undo" action over it.

apeatling commented:
@nylen @jsnajdr Even if this PR removes unused code and matches our current standards, it's objectively a worse experience for our users, and that should be the primary factor.

I'd like to see us try and at the very least provide some visual feedback to the user to help them understand what just happened before we merge this.

jsnajdr commented:
@apeatling Alright, I have some ideas how to implement @shaunandrews's suggestions without any massive rewriting. Let me try them out...

lamosty commented:
Code is looking good. I'll wait for the improvement to UX and then approve. :)

nylen commented:
> One more thing: implementing the notices as global ones (shown in the top right corner) would be much easier from the technical standpoint, although I understand that having them inline is a much neater UX.

The UX will be better for this approach once the global notice has an "Undo" entry.  Otherwise this change could get pretty complicated.

> The main reason for that is: after I trash a Page, it disappears from the "Published" list and moves to the "Trash" one. But the proposal keeps a "shadow" page item in the "Published" list and show a notice with "Undo" action over it.

In the current Redux structure, posts disappear immediately when trashed, via the `trashPost` action creator.  The cleanest way I can think of to change this would be to introduce a new action that does not update optimistically.

a8ci18n commented on client/my-sites/pages/page/index.js:
```diff
@@ -421,42 +410,144 @@ class Page extends Component {
+					},
+					successNotice: {
+						status: 'is-success',
+						text: translate( 'Page restored.' ),
```
Hi! I've found a possible matching string that has already been translated **4** times:
`translate( 'Page restored' )` *ES Score: 16*
See 1 additional suggestion in the [PR translation status page](https://translate.wordpress.com/localci/status/automattic/wp-calypso/remove/update-post-status)

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

a8ci18n commented on client/my-sites/pages/page/index.js:
```diff
@@ -421,42 +410,144 @@ class Page extends Component {
+					progressNotice: {
+						status: 'is-warning',
+						icon: 'history',
+						text: translate( 'Restoring…' ),
```
Hi! I've found a possible matching string that has already been translated **29** times:
`translate( 'Restoring' )` *ES Score: 23*
See 1 additional suggestion in the [PR translation status page](https://translate.wordpress.com/localci/status/automattic/wp-calypso/remove/update-post-status)

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

a8ci18n commented on client/my-sites/pages/page/index.js:
```diff
@@ -421,42 +410,144 @@ class Page extends Component {
+					},
+					successNotice: {
+						status: 'is-success',
+						text: translate( 'Page trashed.' ),
```
Hi! I've found a possible matching string that has already been translated **4** times:
`translate( 'Page trashed' )` *ES Score: 16*

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

a8ci18n commented on client/my-sites/pages/page/index.js:
```diff
@@ -421,42 +410,144 @@ class Page extends Component {
+					progressNotice: {
+						status: 'is-error',
+						icon: 'trash',
+						text: translate( 'Trashing…' ),
```
Hi! I've found a possible matching string that has already been translated **21** times:
`translate( 'Trashing...' )` *ES Score: 22*
See 2 additional suggestions in the [PR translation status page](https://translate.wordpress.com/localci/status/automattic/wp-calypso/remove/update-post-status)

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

a8ci18n commented on client/my-sites/pages/page/index.js:
```diff
@@ -421,42 +410,144 @@ class Page extends Component {
+						},
+						errorNotice: {
+							status: 'is-error',
+							text: translate( 'Failed to delete page.' ),
```
Hi! I've found a possible matching string that has already been translated **101** times:
`translate( 'Failed to delete the page.' )` *ES Score: 12*

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

a8ci18n commented on client/my-sites/pages/page/index.js:
```diff
@@ -421,42 +410,144 @@ class Page extends Component {
+						},
+						successNotice: {
+							status: 'is-success',
+							text: translate( 'Page deleted.' ),
```
Hi! I've found a possible matching string that has already been translated **47** times:
`translate( 'Page Deleted' )` *ES Score: 14*
See 2 additional suggestions in the [PR translation status page](https://translate.wordpress.com/localci/status/automattic/wp-calypso/remove/update-post-status)

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

a8ci18n commented on client/my-sites/pages/page/index.js:
```diff
@@ -421,42 +410,144 @@ class Page extends Component {
+						progressNotice: {
+							status: 'is-error',
+							icon: 'trash',
+							text: translate( 'Deleting…' ),
```
Hi! I've found a possible matching string that has already been translated **31** times:
`translate( 'Deleting' )` *ES Score: 20*
See 2 additional suggestions in the [PR translation status page](https://translate.wordpress.com/localci/status/automattic/wp-calypso/remove/update-post-status)

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

jsnajdr commented:
@apeatling @shaunandrews This branch now has a WIP implementation of progress notices when trashing a site page (or changing the page status in general). Here is how it looks like:

![page-trash-anim](https://user-images.githubusercontent.com/664258/37422204-0ab536a6-27bb-11e8-8bb5-40da18b42279.gif)

It shows a "progress" notice, then a "done" notice, then waits for 5 seconds, then removes the page item from the list completely. There is no "undo" yet, that's a todo.

The main open question is: for how long should the notice with "Page trashed" and the "undo" button be shown? For some limited time? Or forever, until I navigate to another Calypso route?

TODO before this is ready for production:
- implement the Undo action
- add the neat animations that @shaunandrews' mockup shows when the notice appears or changes to something else
- decide what icons and colors should the notices have for all possible actions (page trashed, page restored, page published)
- implement also for the post list (PostTypeList)
- if we decide to remove the "shadow" item (that's how I call items that were removed from the list, but are still shown in the UI) after some time, maybe the removal should be animated?

Test on https://calypso.live/?branch=remove/update-post-status

apeatling commented:
> The main open question is: for how long should the notice with "Page trashed" and the "undo" button be shown? For some limited time? Or forever, until I navigate to another Calypso route?

Can we try leaving it there permanently until you browse to a new route, and see how it feels?

jsnajdr commented:
> Can we try leaving it there permanently until you browse to a new route, and see how it feels?

I pushed a commit where I implemented Undo for trashing and restoring posts. The actions that have Undo available show the success notice permanently (with the 'undo' button), the other ones, which are purely informative, stay for 5 seconds and then disappear.

Please try it out.

a8ci18n commented on client/my-sites/pages/page/index.js:
```diff
@@ -421,42 +435,155 @@ class Page extends Component {
+					progressNotice: {
+						status: 'is-warning',
+						icon: 'history',
+						text: translate( 'Restoring…' ),
```
Hi! I've found a possible matching string that has already been translated **29** times:
`translate( 'Restoring' )` *ES Score: 23*
See 1 additional suggestion in the [PR translation status page](https://translate.wordpress.com/localci/status/automattic/wp-calypso/remove/update-post-status)

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

a8ci18n commented on client/my-sites/pages/page/index.js:
```diff
@@ -421,42 +435,155 @@ class Page extends Component {
+					progressNotice: {
+						status: 'is-error',
+						icon: 'trash',
+						text: translate( 'Trashing…' ),
```
Hi! I've found a possible matching string that has already been translated **21** times:
`translate( 'Trashing...' )` *ES Score: 22*
See 2 additional suggestions in the [PR translation status page](https://translate.wordpress.com/localci/status/automattic/wp-calypso/remove/update-post-status)

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

a8ci18n commented on client/my-sites/pages/page/index.js:
```diff
@@ -421,42 +435,155 @@ class Page extends Component {
+					},
+					errorNotice: {
+						status: 'is-error',
+						text: translate( 'Failed to delete page.' ),
```
Hi! I've found a possible matching string that has already been translated **101** times:
`translate( 'Failed to delete the page.' )` *ES Score: 12*

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

a8ci18n commented on client/my-sites/pages/page/index.js:
```diff
@@ -421,42 +435,155 @@ class Page extends Component {
+					},
+					successNotice: {
+						status: 'is-success',
+						text: translate( 'Page deleted.' ),
```
Hi! I've found a possible matching string that has already been translated **47** times:
`translate( 'Page Deleted' )` *ES Score: 14*
See 2 additional suggestions in the [PR translation status page](https://translate.wordpress.com/localci/status/automattic/wp-calypso/remove/update-post-status)

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

a8ci18n commented on client/my-sites/pages/page/index.js:
```diff
@@ -421,42 +435,155 @@ class Page extends Component {
+					progressNotice: {
+						status: 'is-error',
+						icon: 'trash',
+						text: translate( 'Deleting…' ),
```
Hi! I've found a possible matching string that has already been translated **31** times:
`translate( 'Deleting' )` *ES Score: 20*
See 2 additional suggestions in the [PR translation status page](https://translate.wordpress.com/localci/status/automattic/wp-calypso/remove/update-post-status)

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

apeatling commented:
My internet is not good enough right now to check this. @shaunandrews would you be able to check that this final change makes sense?

nylen commented:
Spotted the inline notice pattern elsewhere (at `/stats/day/:site`):

![2018-03-15t14 31 10-0500](https://user-images.githubusercontent.com/227022/37486852-c14f2f78-285d-11e8-9e73-20726949641e.png)

I'll give this a behavior and code review today.

jsnajdr commented:
> Spotted the inline notice pattern elsewhere (at /stats/day/:site):

That notice has been there just for a few days: https://github.com/Automattic/wp-calypso/pull/23148

@jnylen If you'll be looking at the code, here is a short overview how it works:

The Pages List component gets the array of pages to render from Redux, as the `pages` prop. If I trash a page on the "Published" list, it will immediately disappear from the query result and the Pages List receives a new `pages` prop, without the page I just trashed.

But that's a problem: I want to continue showing the trashed page on the Published list, as what I call a "shadow item". It's grayed out, inactive, and only shows the inline notices with an "undo" action.

I solve that by:
1. Keeping a list of `shadowItems` in Pages List local state. The rendered child `Page`s get a `shadowStatus` prop if they are supposed to be rendered as shadow items.
2. Every time a new `pages` prop is received, I store it in `this.state.pages` and use the list from local state in `render`. The point is that I can choose to NOT receive a `pages` prop update if I don't want to. I do that in case there are any shadow items to display. Then I freeze the `pages` list at a state it was before trashing and ignore the incoming `pages` updates.
3. Only after `shadowItems` becomes empty or when an invalidation event happens (the list was switched to another query, e.g., after switching from "Published" tab to "Drafts"), only then I update the local state with the latest value of the `pages` prop.

I don't know how to implement the same thing in `PostTypeList` yet. As its state is stored in `state.ui.postTypeList`, the above described logic will most likely move to Redux.

samouri commented on client/my-sites/pages/page-list.jsx:
```diff
@@ -154,6 +168,35 @@ class Pages extends Component {
+				// if the last shadow item just got removed, start showing the up-to-date post
+				// list as specified by props.
+				if ( size( newShadowItems ) === 0 ) {
+					newState.pages = this.props.pages;
```
Are you sure it is okay to access `this.props` here? I always figured you were supposed to use the second argument to the setState updater fn.

jsnajdr replied:
Good point, I wasn't aware that the `setState` updater function has a second `props` param. Pushed a fix.

samouri commented on client/my-sites/pages/page/index.js:
```diff
@@ -421,42 +430,151 @@ class Page extends Component {
-					{ this.props.buildUpdateTemplate() }
-				</ReactCSSTransitionGroup>
+				{ ellipsisMenu }
+				{ this.renderShadowNotice() }
```
total nit: don't love functions named `renderX()`.  I usually tend to make a little Component outside of the class definition.

```jsx

function shadowNotice( props ) {
 return ....;
}

class Page extends Component {
  render() {
     { this.props.shadow.shadowStatus && <ShadowNotice { relevantPropsHere } /> }
  }
}
```

jsnajdr replied:
OK, I extracted a `ShadowNotice` component.

samouri commented:
UX Here looks like its gotten pretty good! 
I haven't had the chance to fully wrap my head around the code yet -- the shadow status has a few moving parts.
I'll have time to review it more in depth tomorrow morning

nylen commented:
> But that's a problem: I want to continue showing the trashed page on the Published list, as what I call a "shadow item". It's grayed out, inactive, and only shows the inline notices with an "undo" action.

I think this will overall look much cleaner if we move this logic into Redux...

> As its state is stored in `state.ui.postTypeList`, the above described logic will most likely move to Redux.

... just like you said.  Some rough thoughts on how this would look, in order to be able to integrate with both the page and post lists.  I think the list of "shadow" items should live under `state.ui.shadowPosts` or similar.  It may also be cleaner to remove the "optimistic trashing" logic at [its source](https://github.com/Automattic/wp-calypso/pull/23090/files#diff-9dbc67e30b1741588dd4c9d47fed514cR264), and actually leave the post in its previous status until the API call succeeds.  Then, the logic that renders a page can call `pages: getPostsForQueryIgnoringPage( state, siteId, query )` as it does today, and additionally check each item in that list against `shadowPages: getShadowPostsForQueryIgnoringPage( state, siteId, query)` which contains the information about any in-progress or undo notices.

Both of those things (move this logic to Redux state and make it work for `/posts` as well) are probably larger than necessary for this PR, which seems to be achieving its original goal without affecting UX.

And for an even bigger future task:  Originally we (Tanooki) planned to move `/pages` over to use `PostTypeList`, but that hasn't happened yet due to other priorities.

a8ci18n commented on client/my-sites/pages/page/index.js:
```diff
@@ -421,42 +430,154 @@ class Page extends Component {
+					progressNotice: {
+						status: 'is-warning',
+						icon: 'history',
+						text: translate( 'Restoring…' ),
```
Hi! I've found a possible matching string that has already been translated **29** times:
`translate( 'Restoring' )` *ES Score: 23*
See 1 additional suggestion in the [PR translation status page](https://translate.wordpress.com/localci/status/automattic/wp-calypso/remove/update-post-status)

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

a8ci18n commented on client/my-sites/pages/page/index.js:
```diff
@@ -421,42 +430,154 @@ class Page extends Component {
+					progressNotice: {
+						status: 'is-error',
+						icon: 'trash',
+						text: translate( 'Trashing…' ),
```
Hi! I've found a possible matching string that has already been translated **21** times:
`translate( 'Trashing...' )` *ES Score: 22*
See 2 additional suggestions in the [PR translation status page](https://translate.wordpress.com/localci/status/automattic/wp-calypso/remove/update-post-status)

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

nylen commented on client/my-sites/pages/page/index.js:
```diff
@@ -421,42 +430,154 @@ class Page extends Component {
+		// remove the success/error notice after 5 seconds
+		await sleep( 5000 );
+		await this.changeShadowStatus( false );
+	}
```
I think this async logic should live in an action creator instead of a component, but as I commented previously, that should be fine to happen in a separate PR that also implements this behavior for other post types.

nylen commented on client/state/posts/actions.js:
```diff
@@ -252,7 +258,37 @@ export function savePost( siteId, postId = null, post ) {
+			siteId,
+			postId,
+			post,
+		} );
```
To confirm my understanding here... this is the "optimistic update" that causes the post to immediately disappear from the list, and then some of the "shadow item" code works around that.  But we're not ready to remove the optimistic update because we don't have the in progress and undo UX for the posts list yet.

jsnajdr replied:
Yes. First, I didn't want to make any changes that would affect `PostTypeList` as well. I'm afraid I would have to do a long chain of cascading code updates and never got this PR finished -- it already inflated a lot from the original intent 🙂

Second, I don't want to touch anything related to `PostQueryManager` right now -- that's a component I still don't fully understand and I remember it caused a lot of trouble during the `PostTypeList` project when trying to optimize the Redux dispatches there.

In this PR, I'm looking for a reasonable tradeoff between doing some horrible hack and "doing it properly", which would require rewriting large parts of `PostQueryManager`, `PostTypeList`, the page list UI etc. Let's do that in smaller steps.

Even if the "optimistic update" was removed and Redux state was updated only on `POST_SAVE_SUCCESS`, the "shadow item" logic wouldn't change very much. I still need to continue displaying the item after it's removed from the list, because I need to show the success notice for a few seconds, and the "Undo" action, too.

nylen commented on client/state/posts/actions.js:
```diff
@@ -252,7 +258,37 @@ export function savePost( siteId, postId = null, post ) {
+
+		trashResult
+			.then( savedPost => {
+				dispatch( savePostSuccess( siteId, postId, savedPost, post ) );
```
Why do we need to call `savePostSuccess` here?  It looks like the only thing it does is update the post ID for a newly saved draft from `''` (used by the editor) to the newly created ID, but trashing a post that doesn't actually exist yet shouldn't be possible.

jsnajdr replied:
I moved this part of the changes to a separate PR, because it's fairly independent: #23359. Its description answers most questions.

Here, I'm refactoring the `trashPost` action to use the `wpcom.site().post().delete()` method, but keep its behavior the same as if it was `savePost( { status: 'trash' } )` -- the actions dispatched are the same etc.

This is to avoid too many changes across the app. We don't have any `POST_TRASH_SUCCESS` Redux actions, and there are things that depend on dispatching `POST_SAVE_SUCCESS` after trashing. For example, see [this code that displays a global notice](https://github.com/Automattic/wp-calypso/blob/master/client/state/notices/middleware.js#L131-L146).

nylen replied:
SGTM - I missed some other effects of the `savePostSuccess` action, like sending notices.  I'll look at that PR separately.

nylen commented on client/state/posts/test/actions.js:
```diff
@@ -391,25 +392,29 @@ describe( 'actions', () => {
-					postId: null,
-					error: sinon.match( { message: 'User cannot edit posts' } ),
+			return savePost( 77203074, null, { title: 'Hello World' } )( spy )
+				.catch( noop )
```
Why the `.catch( noop )` here and elsewhere?  If we hit the `catch` block, shouldn't the test actually fail?

jsnajdr replied:
See the description of #23359 for the explanation. The `savePost` dispatch fires the appropriate Redux actions on success and error, and returns the resolved/rejected promise so that the component can update its local state accordingly, too.

What we're testing there is: after the dispatch is finished, no matter if it was success or error, check if the expected actions were dispatched.

The best solution would be `savePost()( spy ).finally( () => { expect( ... ) } )`, but `finally` is very new and is not available everywhere. What is most important, it's not in the version of Node.js we use for running the tests.

nylen commented:
Finally a UX note.  When a page is trashed, it's still possible to click on the edit link.  I think this should be disabled, because all it's going to do is show an error in the editor:

> <img src="https://user-images.githubusercontent.com/227022/37495410-1c61dd94-287b-11e8-8b1c-2320d861a39e.png" width="400">

> <img src="https://user-images.githubusercontent.com/227022/37495446-4d3b6278-287b-11e8-8f83-3f0b62e28df3.png" width="400">

nylen commented:
Similarly, when a trashed page is restored, its URL should update to a valid editor URL:

> ![2018-03-15t18 12 52-0500](https://user-images.githubusercontent.com/227022/37495704-8c3702ba-287c-11e8-8a48-9346dcfeabe0.png)

a8ci18n commented on client/my-sites/pages/page/index.js:
```diff
@@ -421,42 +425,154 @@ class Page extends Component {
+					progressNotice: {
+						status: 'is-warning',
+						icon: 'history',
+						text: translate( 'Restoring…' ),
```
Hi! I've found a possible matching string that has already been translated **29** times:
`translate( 'Restoring' )` *ES Score: 23*
See 1 additional suggestion in the [PR translation status page](https://translate.wordpress.com/localci/status/automattic/wp-calypso/remove/update-post-status)

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

a8ci18n commented on client/my-sites/pages/page/index.js:
```diff
@@ -421,42 +425,154 @@ class Page extends Component {
+					progressNotice: {
+						status: 'is-error',
+						icon: 'trash',
+						text: translate( 'Trashing…' ),
```
Hi! I've found a possible matching string that has already been translated **21** times:
`translate( 'Trashing...' )` *ES Score: 22*
See 2 additional suggestions in the [PR translation status page](https://translate.wordpress.com/localci/status/automattic/wp-calypso/remove/update-post-status)

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

jsnajdr commented:
@nylen I addressed most review issues and pushed a few fixes. `savePostSuccess` and `catch( noop )` have good reasons to be there, and the clickable edit links issue is solved -- a shadow item is now covered with a semi-transparent div and is not clickable.

I couldn't reproduce the issue with wrong edit links. They all seem correct to me.

Regarding the question whether to use Redux for storing the `shadowItems` state: I'm not convinced that Redux is the right approach for things like this or the state stored in `state.ui.postTypeList`. That state is not at all interesting to any other component than `PostTypeList` or `PageList`. It manages UI microinteractions like progress notices or popup menu toggle states. I think that using local state would me much more convenient for that.

For example, `PostItem` and `PostActionsEllipsisMenu` don't need to use `POST_TYPE_LIST_SHARE_PANEL_TOGGLE` and `isSharePanelOpen`: they can use props and local state to communicate the panel toggle flag.

Also, the async logic that shows and hides the shadow notices is tightly bound to UI, and can depend a lot on various CSS animations we might do. What if we used `ReactCSSTransitionGroup` to animate the notices? Then the precise timings in that async logic must be in sync with the animations we do. That logic is not interesting to any other part of the app -- it's local state of the UI. Moving it to Redux doesn't bring any benefits, it only adds coupling between distant parts of the app and more verbosity.

I'd love if we explored various ways of mixing Redux and local state when implementing our UI. For example, there is a new context API coming to React 16.3 and [this article](https://medium.freecodecamp.org/replacing-redux-with-the-new-react-context-api-8f5d01a00e8c) presents a very nice lightweight alternative to Redux that could be used at many places.

shaunandrews commented:
Hope you don't mind; I pushed a small tweak to the overlay color (gray, instead of white) and the entrance animation (less pixels and slightly slower) for the shadow notice.

--

Overall this feels OK. I wish there was some sort of animation when "trashing" or "restoring." The notice alone is OK, but sometime feels "stuck" while you're waiting. If you look at the GIF I posted above, you'll notice the I had the ellipsis is "Trashing..." animate while it was showing. I think something like that for both trashing and restoring states would help a lot with any anxiety about whats happening.

a8ci18n commented on client/my-sites/pages/page/index.js:
```diff
@@ -421,42 +423,154 @@ class Page extends Component {
+					progressNotice: {
+						status: 'is-warning',
+						icon: 'history',
+						text: translate( 'Restoring…' ),
```
Hi! I've found a possible matching string that has already been translated **29** times:
`translate( 'Restoring' )` *ES Score: 23*
See 1 additional suggestion in the [PR translation status page](https://translate.wordpress.com/localci/status/automattic/wp-calypso/remove/update-post-status)

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

a8ci18n commented on client/my-sites/pages/page/index.js:
```diff
@@ -421,42 +423,154 @@ class Page extends Component {
+					progressNotice: {
+						status: 'is-error',
+						icon: 'trash',
+						text: translate( 'Trashing…' ),
```
Hi! I've found a possible matching string that has already been translated **21** times:
`translate( 'Trashing...' )` *ES Score: 22*
See 2 additional suggestions in the [PR translation status page](https://translate.wordpress.com/localci/status/automattic/wp-calypso/remove/update-post-status)

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

jsnajdr commented:
> I wish there was some sort of animation when "trashing" or "restoring."

Animating the three dots is not hard in the EN locale, but I'm worried about other languages. When translated to Japanese, for example, the ellipsis in "Trashing…" will be a different Unicode character.

Is there any other way how to animate the notice that's more i18n friendly?

shaunandrews commented:
> Animating the three dots is not hard in the EN locale, but I'm worried about other languages.

Ah, good point. Maybe we take a cue from our placeholder loaders, and do a fade-in-out animation while trashing/restoring on the icon background:

![loading notice](https://user-images.githubusercontent.com/191598/37602326-7d102418-2b62-11e8-9b12-dd8c40ed7741.gif)

nylen commented:
The behavior for trashing a page in the list is working well.

When I restore a page from the Trashed list, I'd expect to be able to click the title to edit it after it's restored.  The new transparent overlay doesn't allow this.

This is probably ok, we're getting into edge case territory there.  But if we're going to fix that (allow clicking to edit a restored page in the Trashed list), then the edit link will still need to be updated - currently it's created as though the restored page is still Trashed:

![2018-03-19t11 56 38-0500](https://user-images.githubusercontent.com/227022/37610033-99e84ed0-2b6c-11e8-80f6-2e1bdc9cc758.png)

nylen commented:
> Regarding the question whether to use Redux for storing the `shadowItems` state: I'm not convinced that Redux is the right approach for things like this or the state stored in `state.ui.postTypeList`.

In the `PostTypeList` there are some subtle changes that need to happen in other components when bulk edit mode is enabled (hide the search icon for example).  This could probably be done using React state and props, but if I recall correctly, we implemented this in Redux state because we wanted to avoid passing down chains of props through multiple components, and we anticipated more functionality that would not be available in bulk edit mode.

In this case, with "shadow post items" I imagine it working mostly using typical Redux actions.  Something like this:

- We remove the optimistic update behavior so that trashed pages/posts do not immediately appear as trashed in the application state.
- The lists of pages and posts query Redux state for `isTrashingPost` and use this to display the in-progress overlay.  This part maps especially well to our existing practices for storing application state.
- After the trash operation completes, we could display a global notice, which will keep the code simpler.  If we really want the UX of the trashed/restored item staying in the list, we need "shadow items" logic like in this PR, but I think the code would probably still be simpler since we are able to handle part of this UX using the normal application state.

In this PR, the logic around in-progress and undo states is implemented completely in local state, which necessitates some new code:

> The `savePost` dispatch fires the appropriate Redux actions on success and error, and returns the resolved/rejected promise so that the component can update its local state accordingly, too.

I think this duplication of information where the same event is handled in multiple places (in Redux action creators and in local component state) is worth avoiding.  Over time this will lead to branching logic with side effects in lots of different places, and I think one of the benefits of Redux is that (in theory, at least) all of this event handling happens in a single type of code.  In practice pure Redux side effects already get pretty hard to follow, like how I missed the point of the `savePostSuccess` call in this PR.

nylen commented:
^ Having said all of that, I think now is not the time to continue trying to simplify this behavior.  A better opportunity would be when we implement the in-progress and undo functionality for posts and other post types.  I'm comfortable with this change once the Restore behavior is fixed.

jsnajdr commented:
@nylen Showing the correct edit link on restored post and making it clickable (i.e., not covering it with the semi-transparent overlay) is a bit more tricky than I expected. I'll implement that as a standalone PR.

I'll do the same with the fade-in-out animation on the notice suggested by @shaunandrews.
